posts/2022-06-29-fixing-haskell-filepaths.md

```diff
@@ -0,0 +1,329 @@
+3. since Haskell `Char` also includes surrogates, the conversion from `String` to e.g. UTF-8 `CString` can in fact fail, so is not total
+
+I have assembled a [list of correctness issues](https://gist.github.com/hasufell/c600d318bdbe010a7841cc351c835f92)
+with these approaches for in-depth reading.
+
+### Why do we care?
+
+Filepaths are resources on the (users) system. We create, delete, copy them. Any corner case with filepaths can have devastating effects: deleting the wrong file, comparing the wrong files, failing whitelists, security bugs, etc.
+
+## The solution
+
+Just stop converting filepaths!
+
+We can just keep the original bytes from the system API. Many filepath operations actually don't need to know the exact underlying encoding.
+E.g. the filepath separator `/` on unix is a pre-defined *byte* (`0x2F`). You can just scan the byte array for this byte.
+The position doesn't matter, the encoding doesn't matter. File **names** *cannot* include this byte, period.
+
+However, since unix and windows are different (`[Word8]` vs `[Word16]`), any API that deals with low-level filepaths in a
+cross-platform manner needs to understand this and write correct code. More on this in the migration strategy section below.
+
+We decided to use `ShortByteString` as the internal representation of filepaths, because:
+
+1. these are raw, uninterpreted bytes, a wrapper around `ByteArray#`, which has many efficient primops
+2. it's unpinned, so doesn't contribute to memory fragmentation ([proof](https://github.com/hasufell/filepath-debug/blob/master/result.txt))
+3. provides convenient API via `bytestring`, which has been [greatly enhanced as part of this proposal](https://github.com/haskell/bytestring/pull/471)
+
+## Migration strategies
+
+Most end-users developing applications should be able to convert to the new API with little effort, given that their favorite libraries
+already support this new type.
+
+[System.OsPath](https://hackage.haskell.org/package/filepath-1.4.99.0/candidate/docs/System-OsPath.html) exports the same API as `System.FilePath` with some additional helpers to convert from and to `String`. Likewise `System.OsPath.Posix`/`System.OsPath.Windows` are equivalent to `System.FilePath.Posix`/`System.FilePath.Windows`.
+
+So, you can just:
+
+1. update your dependencies lower bounds to the minimum version that supports `OsPath`
+2. import `System.OsPath` instead of `System.FilePath`
+3. use the specialised API from your dependencies, e.g. for unix `System.Posix.Directory.PosixPath` instead of `System.Posix.Directory`
+4. there's currently no `fromString` instance, so `OverloadedStrings` doesn't work (there's no total instance, similar to [Surprising behavior of ByteString literals via IsString](https://github.com/haskell/bytestring/issues/140)). Instead, we provide QuasiQuoters. If you're just using an ASCII subset, you can use `System.OsPath.encodeUtf` and `System.OsPath.decodeUtf` with `fromJust`.
```

Comment:
Some beginner-friendly documentation would be helpful here. A simple illustration of how to implement "open a file at /tmp/foo.txt and read it" would clarify a lot of things. Beginners won't know what QuasiQuoters are or how to use them.

Reply:
```hs
{-# LANGUAGE QuasiQuotes #-}

import qualified System.File.OsPath as OSP
import               System.OsPath

main :: IO ()
main = do
  OSP.readFile [osp|/etc/passwd|]
```

Reply:
Thanks, that's a nice simple demo. Would you mind adding it to the article?